File: next/scripts/validate_i18n_messages.py
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
MESSAGES_DIR = ROOT / "apps" / "web" / "messages"
LOCALES_TS = ROOT / "apps" / "web" / "i18n" / "locales.ts"
BASE_LOCALE = "en"

PLACEHOLDER_RE = re.compile(r"\{[^{}]+\}")


def fail(message: str) -> None:
    print(f"[i18n:validate] ERROR: {message}")
    sys.exit(1)


def parse_supported_locales() -> list[str]:
    content = LOCALES_TS.read_text(encoding="utf-8")
    match = re.search(r"SUPPORTED_LOCALES\s*=\s*\[(.*?)\]\s*as\s*const", content, re.S)
    if not match:
        fail("Impossibile trovare SUPPORTED_LOCALES in i18n/locales.ts")
    chunk = match.group(1)
    return re.findall(r'"([A-Za-z_]+)"', chunk)


def flatten(node: object, prefix: str = "") -> dict[str, str]:
    out: dict[str, str] = {}
    if isinstance(node, dict):
        for key, value in node.items():
            child_prefix = f"{prefix}.{key}" if prefix else key
            out.update(flatten(value, child_prefix))
    elif isinstance(node, str):
        out[prefix] = node
    else:
        # In questo progetto i messaggi sono stringhe; altri tipi non sono ammessi.
        fail(f"Valore non-stringa trovato in chiave '{prefix}': {type(node).__name__}")
    return out


def placeholders(text: str) -> tuple[str, ...]:
    return tuple(sorted(PLACEHOLDER_RE.findall(text)))


def main() -> None:
    locales = parse_supported_locales()

    missing_files = [loc for loc in locales if not (MESSAGES_DIR / f"{loc}.json").exists()]
    if missing_files:
        fail(f"File lingua mancanti: {', '.join(missing_files)}")

    extra_files = sorted(
        p.stem
        for p in MESSAGES_DIR.glob("*.json")
        if p.stem not in locales
    )
    if extra_files:
        fail(f"File lingua non dichiarati in SUPPORTED_LOCALES: {', '.join(extra_files)}")

    base_path = MESSAGES_DIR / f"{BASE_LOCALE}.json"
    if not base_path.exists():
        fail("File base en.json non trovato")

    base_obj = json.loads(base_path.read_text(encoding="utf-8"))
    base_flat = flatten(base_obj)
    base_keys = set(base_flat)

    errors: list[str] = []

    for locale in locales:
        path = MESSAGES_DIR / f"{locale}.json"
        try:
            obj = json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError as exc:
            errors.append(f"{locale}: JSON non valido ({exc})")
            continue

        flat = flatten(obj)
        keys = set(flat)

        missing = sorted(base_keys - keys)
        extra = sorted(keys - base_keys)
        if missing:
            errors.append(f"{locale}: chiavi mancanti ({len(missing)}), esempio: {', '.join(missing[:5])}")
        if extra:
            errors.append(f"{locale}: chiavi extra ({len(extra)}), esempio: {', '.join(extra[:5])}")

        # Confronto placeholder per tutte le chiavi comuni
        for key in sorted(base_keys & keys):
            base_ph = placeholders(base_flat[key])
            loc_ph = placeholders(flat[key])
            if base_ph != loc_ph:
                errors.append(
                    f"{locale}: placeholder diversi in '{key}' (base={base_ph}, locale={loc_ph})"
                )

    if errors:
        print("[i18n:validate] ERRORI trovati:")
        for err in errors:
            print(f"- {err}")
        sys.exit(1)

    print(f"[i18n:validate] OK: {len(locales)} locale validati, schema e placeholder coerenti con {BASE_LOCALE}.json")


if __name__ == "__main__":
    main()

File: next/scripts/validate-i18n-messages.mjs
import { readdir, readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const MESSAGES_DIR = path.join(ROOT, "apps", "web", "messages");
const LOCALES_TS = path.join(ROOT, "apps", "web", "i18n", "locales.ts");
const BASE_LOCALE = "en";
const PLACEHOLDER_RE = /\{[^{}]+\}/g;

async function main() {
  const locales = await parseSupportedLocales();
  const messageFiles = await readdir(MESSAGES_DIR);
  const jsonLocales = messageFiles
    .filter((file) => file.endsWith(".json"))
    .map((file) => path.basename(file, ".json"))
    .sort();

  const missingFiles = locales.filter((locale) => !jsonLocales.includes(locale));
  if (missingFiles.length > 0) {
    fail(`Missing locale file(s): ${missingFiles.join(", ")}`);
  }

  const extraFiles = jsonLocales.filter((locale) => !locales.includes(locale));
  if (extraFiles.length > 0) {
    fail(`Locale file(s) not declared in SUPPORTED_LOCALES: ${extraFiles.join(", ")}`);
  }

  const basePath = path.join(MESSAGES_DIR, `${BASE_LOCALE}.json`);
  const baseObject = await readJson(basePath, BASE_LOCALE);
  const baseFlat = flatten(baseObject);
  const baseKeys = new Set(Object.keys(baseFlat));
  const errors = [];

  for (const locale of locales) {
    const filePath = path.join(MESSAGES_DIR, `${locale}.json`);
    const messages = await readJson(filePath, locale, errors);
    if (!messages) continue;

    const flat = flatten(messages, "", errors);
    const keys = new Set(Object.keys(flat));
    const missing = [...baseKeys].filter((key) => !keys.has(key)).sort();
    const extra = [...keys].filter((key) => !baseKeys.has(key)).sort();

    if (missing.length > 0) {
      errors.push(`${locale}: missing key(s) (${missing.length}), example: ${missing.slice(0, 5).join(", ")}`);
    }
    if (extra.length > 0) {
      errors.push(`${locale}: extra key(s) (${extra.length}), example: ${extra.slice(0, 5).join(", ")}`);
    }

    for (const key of [...baseKeys].filter((entry) => keys.has(entry)).sort()) {
      const basePlaceholders = placeholders(baseFlat[key]);
      const localePlaceholders = placeholders(flat[key]);
      if (!sameArray(basePlaceholders, localePlaceholders)) {
        errors.push(
          `${locale}: placeholder mismatch in '${key}' (base=${formatArray(basePlaceholders)}, locale=${formatArray(localePlaceholders)})`,
        );
      }
    }
  }

  if (errors.length > 0) {
    console.error("[i18n:validate] Errors found:");
    for (const error of errors) {
      console.error(`- ${error}`);
    }
    process.exit(1);
  }

  console.log(`[i18n:validate] OK: validated ${locales.length} locale(s); schema and placeholders match ${BASE_LOCALE}.json`);
}

async function parseSupportedLocales() {
  const content = await readFile(LOCALES_TS, "utf8");
  const match = content.match(/SUPPORTED_LOCALES\s*=\s*\[(.*?)\]\s*as\s*const/s);
  if (!match) {
    fail("Unable to find SUPPORTED_LOCALES in i18n/locales.ts");
  }

  return [...match[1].matchAll(/"([A-Za-z_]+)"/g)].map((entry) => entry[1]);
}

async function readJson(filePath, locale, errors = null) {
  try {
    return JSON.parse(await readFile(filePath, "utf8"));
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    if (errors) {
      errors.push(`${locale}: invalid JSON (${message})`);
      return null;
    }
    fail(`${locale}: invalid JSON (${message})`);
  }
}

function flatten(node, prefix = "", errors = null) {
  const result = {};

  if (node && typeof node === "object" && !Array.isArray(node)) {
    for (const [key, value] of Object.entries(node)) {
      const childPrefix = prefix ? `${prefix}.${key}` : key;
      Object.assign(result, flatten(value, childPrefix, errors));
    }
    return result;
  }

  if (typeof node === "string") {
    result[prefix] = node;
    return result;
  }

  const message = `Non-string value found at key '${prefix}': ${Array.isArray(node) ? "array" : typeof node}`;
  if (errors) {
    errors.push(message);
    return result;
  }
  fail(message);
}

function placeholders(text) {
  return [...text.matchAll(PLACEHOLDER_RE)].map((match) => match[0]).sort();
}

function sameArray(left, right) {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function formatArray(values) {
  return `[${values.join(", ")}]`;
}

function fail(message) {
  console.error(`[i18n:validate] ERROR: ${message}`);
  process.exit(1);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
